crates/sui-types/src/programmable_transaction_builder.rs

```diff
@@ -167,7 +169,15 @@ impl ProgrammableTransactionBuilder {
     }
 
     pub fn publish(&mut self, modules: Vec<Vec<u8>>) {
-        self.commands.push(Command::Publish(modules))
+        let cap = self.command(Command::Publish(modules));
```

Comment:
publish + make immutable feels like the most logical choice for easy migration. 

But thoughts on something like `publish_and_take_cap`, which transfers it to the sender. But that got me thinking... maybe we need something like `Argument::Sender` and maybe `Argument::Sponsor`? Since it's actually quite difficult to add something that transfers the publish cap to the sender, without hard coding the sender address as an argument... which would also needlessly increase the txn size. 

Thoughts?

Reply:
+1 `Argument::sender` at least -- that would help a lot. cc @tzakian as this will be relevant to you creating a test that sends the cap back to sender!

Reply:
Bit of a drive-by comment, but I was just looking at this exact piece of code and thinking the exact same thing you are with having a `publish_and_take_cap`. (for context: I'm working on enabling grabbing the UpgradeCap in transactional tests ATM). I was running into the exact issue you're mentioning of this would then lead to hard-coding the sender address in the `Command` which isn't all that great..

Having `Argument::Sender` (at least) would make sense to me, and `Sponsor` to some extent (although not as much as `Sender` IMO).

Reply:
Great minds!